pykafka/producer.py

```diff
@@ -242,33 +239,54 @@ def produce(self, message, partition_key=None):
             raise ProducerStoppedException()
         partitions = list(self._topic.partitions.values())
         partition_id = self._partitioner(partitions, partition_key).id
-        message_partition_tup = (partition_key, message), partition_id, 0
-        self._produce(message_partition_tup)
+
+        msg = Message(value=message,
+                      partition_key=partition_key,
+                      partition_id=partition_id,
+                      # We must pass our thread-local Queue instance directly,
+                      # as results will be written to it in a worker thread
+                      delivery_report_q=self._delivery_reports.queue)
+        self._produce(msg)
+
         if self._synchronous:
-            self._wait_all()
+            reported_msg, exc = self.get_delivery_report()
+            assert reported_msg is msg
+            if exc is not None:
+                raise exc
         self._raise_worker_exceptions()
 
-    def _produce(self, message_partition_tup):
+    def get_delivery_report(self, block=True, timeout=None):
+        """Fetch delivery reports for messages produced on the current thread
+
+        Returns 2-tuples of a `pykafka.protocol.Message` and either `None`
+        (for successful deliveries) or `Exception` (for failed deliveries)
+        """
+        return self._delivery_reports.queue.get(block, timeout)
+
+    def _produce(self, message):
         """Enqueue a message for the relevant broker
 
-        :param message_partition_tup: Message with partition assigned.
-        :type message_partition_tup: ((bytes, bytes), int) tuple
+        :param message: Message with valid `partition_id`, ready to be sent
+        :type message: `pykafka.protocol.Message`
         """
-        kv, partition_id, attempts = message_partition_tup
         success = False
         while not success:
-            leader_id = self._topic.partitions[partition_id].leader.id
+            leader_id = self._topic.partitions[message.partition_id].leader.id
             if leader_id in self._owned_brokers:
-                self._owned_brokers[leader_id].enqueue([(kv, partition_id, attempts)])
+                self._owned_brokers[leader_id].enqueue(message)
                 success = True
             else:
                 success = False
 
+    def _put_delivery_report(self, msg, exc=None):
```

Comment:
It might grow to a two-liner when we check if reporting is enabled or not, but yeah, it could still be quickly factored out. I had mostly been thinking of what it might all look like if we ended up with a sort of mixin :)